### PR TITLE
server model

### PR DESCRIPTION
**Pull request by julienledem**
Signed-off-by: Julien Le Dem <julien@apache.org>

### CODE REVIEW

**Comment by julienledem** [comment]
the new classes look like this:
```
package io.openlineage.server;

import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import java.lang.Object;
import java.lang.String;
import java.net.URI;
import java.time.ZonedDateTime;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public final class OpenLineage {
  @JsonPropertyOrder
  public static final class RunFacets {
    @JsonAnySetter
    private final Map<String, RunFacet> additionalProperties;

    @JsonCreator
    private RunFacets() {
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, RunFacet> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "eventType",
      "eventTime",
      "run",
      "job",
      "inputs",
      "outputs",
      "producer",
      "schemaURL"
  })
  public static final class RunEvent {
    private final String eventType;

    private final ZonedDateTime eventTime;

    private final Run run;

    private final Job job;

    private final List<InputDataset> inputs;

    private final List<OutputDataset> outputs;

    private final URI producer;

    private final URI schemaURL;

    /**
     * @param eventType the current transition of the run state. It is required to issue 1 START event and 1 of [ COMPLETE, ABORT, FAIL ] event per run. Additional events with OTHER eventType can be added to the same run. For example to send additional metadata after the run is complete
     * @param eventTime the time the event occurred at
     * @param run the run
     * @param job the job
     * @param inputs The set of **input** datasets.
     * @param outputs The set of **output** datasets.
     * @param producer URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     * @param schemaURL The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this RunEvent
     */
    @JsonCreator
    private RunEvent(@JsonProperty("eventType") String eventType,
        @JsonProperty("eventTime") ZonedDateTime eventTime, @JsonProperty("run") Run run,
        @JsonProperty("job") Job job, @JsonProperty("inputs") List<InputDataset> inputs,
        @JsonProperty("outputs") List<OutputDataset> outputs,
        @JsonProperty("producer") URI producer, @JsonProperty("schemaURL") URI schemaURL) {
      this.eventType = eventType;
      this.eventTime = eventTime;
      this.run = run;
      this.job = job;
      this.inputs = inputs;
      this.outputs = outputs;
      this.producer = producer;
      this.schemaURL = schemaURL;
    }

    /**
     * @return the current transition of the run state. It is required to issue 1 START event and 1 of [ COMPLETE, ABORT, FAIL ] event per run. Additional events with OTHER eventType can be added to the same run. For example to send additional metadata after the run is complete
     */
    public String getEventType() {
      return eventType;
    }

    /**
     * @return the time the event occurred at
     */
    public ZonedDateTime getEventTime() {
      return eventTime;
    }

    public Run getRun() {
      return run;
    }

    public Job getJob() {
      return job;
    }

    /**
     * @return The set of **input** datasets.
     */
    public List<InputDataset> getInputs() {
      return inputs;
    }

    /**
     * @return The set of **output** datasets.
     */
    public List<OutputDataset> getOutputs() {
      return outputs;
    }

    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    public URI getProducer() {
      return producer;
    }

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this RunEvent
     */
    public URI getSchemaURL() {
      return schemaURL;
    }
  }

  @JsonPropertyOrder
  public static final class JobFacets {
    @JsonAnySetter
    private final Map<String, JobFacet> additionalProperties;

    @JsonCreator
    private JobFacets() {
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, JobFacet> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder
  public static final class InputDatasetInputFacets {
    @JsonAnySetter
    private final Map<String, InputDatasetFacet> additionalProperties;

    @JsonCreator
    private InputDatasetInputFacets() {
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, InputDatasetFacet> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "_producer",
      "_schemaURL"
  })
  public static final class DatasetFacet implements BaseFacet {
    private final URI _producer;

    private final URI _schemaURL;

    @JsonAnySetter
    private final Map<String, Object> additionalProperties;

    /**
     * @param _producer URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     * @param _schemaURL The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    @JsonCreator
    private DatasetFacet(@JsonProperty("_producer") URI _producer,
        @JsonProperty("_schemaURL") URI _schemaURL) {
      this._producer = _producer;
      this._schemaURL = _schemaURL;
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    public URI get_producer() {
      return _producer;
    }

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    public URI get_schemaURL() {
      return _schemaURL;
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, Object> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "_producer",
      "_schemaURL"
  })
  public static final class OutputDatasetFacet implements BaseFacet {
    private final URI _producer;

    private final URI _schemaURL;

    @JsonAnySetter
    private final Map<String, Object> additionalProperties;

    /**
     * @param _producer URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     * @param _schemaURL The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    @JsonCreator
    private OutputDatasetFacet(@JsonProperty("_producer") URI _producer,
        @JsonProperty("_schemaURL") URI _schemaURL) {
      this._producer = _producer;
      this._schemaURL = _schemaURL;
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    public URI get_producer() {
      return _producer;
    }

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    public URI get_schemaURL() {
      return _schemaURL;
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, Object> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder
  public static final class DatasetFacets {
    @JsonAnySetter
    private final Map<String, DatasetFacet> additionalProperties;

    @JsonCreator
    private DatasetFacets() {
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, DatasetFacet> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "runId",
      "facets"
  })
  public static final class Run {
    private final UUID runId;

    private final RunFacets facets;

    /**
     * @param runId The globally unique ID of the run associated with the job.
     * @param facets The run facets.
     */
    @JsonCreator
    private Run(@JsonProperty("runId") UUID runId, @JsonProperty("facets") RunFacets facets) {
      this.runId = runId;
      this.facets = facets;
    }

    /**
     * @return The globally unique ID of the run associated with the job.
     */
    public UUID getRunId() {
      return runId;
    }

    /**
     * @return The run facets.
     */
    public RunFacets getFacets() {
      return facets;
    }
  }

  public interface Dataset {
    /**
     * @return The namespace containing that dataset
     */
    String getNamespace();

    /**
     * @return The unique name for that dataset within that namespace
     */
    String getName();

    /**
     * @return The facets for this dataset
     */
    DatasetFacets getFacets();
  }

  @JsonPropertyOrder
  public static final class OutputDatasetOutputFacets {
    @JsonAnySetter
    private final Map<String, OutputDatasetFacet> additionalProperties;

    @JsonCreator
    private OutputDatasetOutputFacets() {
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, OutputDatasetFacet> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "_producer",
      "_schemaURL"
  })
  public static final class RunFacet implements BaseFacet {
    private final URI _producer;

    private final URI _schemaURL;

    @JsonAnySetter
    private final Map<String, Object> additionalProperties;

    /**
     * @param _producer URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     * @param _schemaURL The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    @JsonCreator
    private RunFacet(@JsonProperty("_producer") URI _producer,
        @JsonProperty("_schemaURL") URI _schemaURL) {
      this._producer = _producer;
      this._schemaURL = _schemaURL;
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    public URI get_producer() {
      return _producer;
    }

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    public URI get_schemaURL() {
      return _schemaURL;
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, Object> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "namespace",
      "name",
      "facets",
      "outputFacets"
  })
  public static final class OutputDataset implements Dataset {
    private final String namespace;

    private final String name;

    private final DatasetFacets facets;

    private final OutputDatasetOutputFacets outputFacets;

    /**
     * @param namespace The namespace containing that dataset
     * @param name The unique name for that dataset within that namespace
     * @param facets The facets for this dataset
     * @param outputFacets The output facets for this dataset
     */
    @JsonCreator
    private OutputDataset(@JsonProperty("namespace") String namespace,
        @JsonProperty("name") String name, @JsonProperty("facets") DatasetFacets facets,
        @JsonProperty("outputFacets") OutputDatasetOutputFacets outputFacets) {
      this.namespace = namespace;
      this.name = name;
      this.facets = facets;
      this.outputFacets = outputFacets;
    }

    /**
     * @return The namespace containing that dataset
     */
    public String getNamespace() {
      return namespace;
    }

    /**
     * @return The unique name for that dataset within that namespace
     */
    public String getName() {
      return name;
    }

    /**
     * @return The facets for this dataset
     */
    public DatasetFacets getFacets() {
      return facets;
    }

    /**
     * @return The output facets for this dataset
     */
    public OutputDatasetOutputFacets getOutputFacets() {
      return outputFacets;
    }
  }

  public interface BaseFacet {
    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    URI get_producer();

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    URI get_schemaURL();

    /**
     * @return additional properties
     */
    Map<String, Object> getAdditionalProperties();
  }

  @JsonPropertyOrder({
      "namespace",
      "name",
      "facets",
      "inputFacets"
  })
  public static final class InputDataset implements Dataset {
    private final String namespace;

    private final String name;

    private final DatasetFacets facets;

    private final InputDatasetInputFacets inputFacets;

    /**
     * @param namespace The namespace containing that dataset
     * @param name The unique name for that dataset within that namespace
     * @param facets The facets for this dataset
     * @param inputFacets The input facets for this dataset.
     */
    @JsonCreator
    private InputDataset(@JsonProperty("namespace") String namespace,
        @JsonProperty("name") String name, @JsonProperty("facets") DatasetFacets facets,
        @JsonProperty("inputFacets") InputDatasetInputFacets inputFacets) {
      this.namespace = namespace;
      this.name = name;
      this.facets = facets;
      this.inputFacets = inputFacets;
    }

    /**
     * @return The namespace containing that dataset
     */
    public String getNamespace() {
      return namespace;
    }

    /**
     * @return The unique name for that dataset within that namespace
     */
    public String getName() {
      return name;
    }

    /**
     * @return The facets for this dataset
     */
    public DatasetFacets getFacets() {
      return facets;
    }

    /**
     * @return The input facets for this dataset.
     */
    public InputDatasetInputFacets getInputFacets() {
      return inputFacets;
    }
  }

  @JsonPropertyOrder({
      "namespace",
      "name",
      "facets"
  })
  public static final class Job {
    private final String namespace;

    private final String name;

    private final JobFacets facets;

    /**
     * @param namespace The namespace containing that job
     * @param name The unique name for that job within that namespace
     * @param facets The job facets.
     */
    @JsonCreator
    private Job(@JsonProperty("namespace") String namespace, @JsonProperty("name") String name,
        @JsonProperty("facets") JobFacets facets) {
      this.namespace = namespace;
      this.name = name;
      this.facets = facets;
    }

    /**
     * @return The namespace containing that job
     */
    public String getNamespace() {
      return namespace;
    }

    /**
     * @return The unique name for that job within that namespace
     */
    public String getName() {
      return name;
    }

    /**
     * @return The job facets.
     */
    public JobFacets getFacets() {
      return facets;
    }
  }

  @JsonPropertyOrder({
      "_producer",
      "_schemaURL"
  })
  public static final class InputDatasetFacet implements BaseFacet {
    private final URI _producer;

    private final URI _schemaURL;

    @JsonAnySetter
    private final Map<String, Object> additionalProperties;

    /**
     * @param _producer URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     * @param _schemaURL The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    @JsonCreator
    private InputDatasetFacet(@JsonProperty("_producer") URI _producer,
        @JsonProperty("_schemaURL") URI _schemaURL) {
      this._producer = _producer;
      this._schemaURL = _schemaURL;
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    public URI get_producer() {
      return _producer;
    }

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    public URI get_schemaURL() {
      return _schemaURL;
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, Object> getAdditionalProperties() {
      return additionalProperties;
    }
  }

  @JsonPropertyOrder({
      "_producer",
      "_schemaURL"
  })
  public static final class JobFacet implements BaseFacet {
    private final URI _producer;

    private final URI _schemaURL;

    @JsonAnySetter
    private final Map<String, Object> additionalProperties;

    /**
     * @param _producer URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     * @param _schemaURL The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    @JsonCreator
    private JobFacet(@JsonProperty("_producer") URI _producer,
        @JsonProperty("_schemaURL") URI _schemaURL) {
      this._producer = _producer;
      this._schemaURL = _schemaURL;
      this.additionalProperties = new LinkedHashMap<>();
    }

    /**
     * @return URI identifying the producer of this metadata. For example this could be a git url with a given tag or sha
     */
    public URI get_producer() {
      return _producer;
    }

    /**
     * @return The JSON Pointer (https://tools.ietf.org/html/rfc6901) URL to the corresponding version of the schema definition for this facet
     */
    public URI get_schemaURL() {
      return _schemaURL;
    }

    /**
     * @return additional properties
     */
    @JsonAnyGetter
    public Map<String, Object> getAdditionalProperties() {
      return additionalProperties;
    }
  }
}
```

File: client/java/src/test/java/io/openlineage/client/OpenLineageTest.java [hunk 1]
```diff
@@ -145,42 +145,56 @@ public void factory() throws JsonProcessingException {
...
+      OutputDataset outputDataset = runStateUpdate.getOutputs().get(0);
+      assertEquals("ons", outputDataset.getNamespace());
+      assertEquals("output", outputDataset.getName());
+
+      assertEquals(roundTrip(json), roundTrip(mapper.writeValueAsString(read)));
+      assertEquals((Long)10L, outputDataset.getOutputFacets().getOutputStatistics().getRowCount());
+      assertEquals((Long)20L, outputDataset.getOutputFacets().getOutputStatistics().getSize());
+
+      assertEquals(json, mapper.writeValueAsString(read));
+    }
```
Comment (julienledem) [inline client/java/src/test/java/io/openlineage/client/OpenLineageTest.java:184]: this part has not changed, I just isolated the variable scope.

**Comment by collado-mike** [comment]
This may be a big ask, but given that the server model classes are all static with public constructors, the enclosing `OpenLineage` class is really just a namespace and will never be instantiated. Is it very difficult to unnest the enclosed classes and make them all top level? It's not the end of the world if it is, but I thought I'd ask.

File: client/java/generator/src/main/java/io/openlineage/client/Generator.java [hunk 2]
```diff
@@ -61,7 +64,8 @@ public static void main(String[] args) throws JsonParseException, JsonMappingExc
       }
     }
     logger.info("Generating code for schemas:\n" + urls.stream().map(Object::toString).collect(Collectors.joining("\n")));
-    generate(urls, new File("src/main/java/io/openlineage/client/"));
+    generate(new HashSet<>(asList(urls.iterator().next())), "io.openlineage.server", true, new File("src/main/java/io/openlineage/server/"));
```
Comment (wslulciuc) [inline client/java/generator/src/main/java/io/openlineage/client/Generator.java:67]: > `io.openlineage.server`

Nice idea to differentiate between `client` and `server` models 💯

File: client/java/src/test/java/io/openlineage/client/OpenLineageTest.java (hunk 1 above)
Comment (wslulciuc) [reply client/java/src/test/java/io/openlineage/client/OpenLineageTest.java:184]: 👍

**Comment by wslulciuc** [comment]
> Is it very difficult to unnest the enclosed classes and make them all top level?

@collado-mike: though I do agree, I'd rather be consistent with the way the `client` models are generated and use the enclosed classes for the `server` as well:

```
io.openlineage.server.OpenLineage.*
io.openlineage.client.OpenLineage.*
```

In a way, the `OpenLineage` is effectively the `models` path that you'd normally find in clients, for example [`marquez.client.models`](https://github.com/MarquezProject/marquez/tree/main/clients/java/src/main/java/marquez/client/models)

**Comment by julienledem** [comment]
I would not change the nesting on the client side now since it breaks compatibility.
I can look into splitting that on the server side but it will make the generation a bit more complex to handle both cases.
I think it is a bit a matter of taste as they are strictly equivalent functionally.

**Comment by collado-mike** [comment]
> @collado-mike: though I do agree, I'd rather be consistent with the way the `client` models are generated and use the enclosed classes for the `server` as well:

I think the client models are the way they are for a reason- trying to make it easy for users to pass in the `_producer` field by hiding all the constructors and using factory methods that all have access to the common `_producer`. The server model has no such need, as we're just deserializing received models.

On the other hand, it is very _inconsistent_ with how Java models are typically written, e.g., in Marquez and other projects. It really is just an aesthetic concern- they all end up in their own `.class` files in the end- but I'm just hoping to make it a bit easier for readers and users to browse the code in a format that is more familiar, thus easier to consume and understand.

**Comment by collado-mike** [comment]
> I would not change the nesting on the client side now since it breaks compatibility.
> I can look into splitting that on the server side but it will make the generation a bit more complex to handle both cases.
> I think it is a bit a matter of taste as they are strictly equivalent functionally.

I agree on all points. Again, I don't want to make a ton of work or to make the generator way more complex than necessary. I was just hoping to make the code a bit more typical of a Java codebase.

**Comment by wslulciuc** [comment]
> I'm just hoping to make it a bit easier for readers and users to browse the code in a format that is more familiar, thus easier to consume and understand

Yeah, I agree 👍 Maybe `server` and `client` pkg naming don't have to be aligned, but I guess I tend to favor consistency but in this case it comes at a cost of usability.

**Review by wslulciuc — APPROVED** [review]
LGTM 👍

**Comment by wslulciuc** [comment]
@julienledem: Are the serve models publish? I don't see the server package on [`o.openlineage.server`](https://openlineage.io/docs/javadoc/io/openlineage/server/package-summary.html) as [`io.openlineage.client`](https://openlineage.io/docs/javadoc/io/openlineage/client/package-summary.html). Let me know if I'm looking in the wrong place.

**Pull request merged**